rotkehlchen/chain/evm/decoding/uniswap_v2_like/decoder.py

```diff
@@ -0,0 +1,140 @@
+import logging
+from collections.abc import Callable
+from typing import TYPE_CHECKING, Literal
+
+from rotkehlchen.assets.asset import EvmToken
+from rotkehlchen.chain.ethereum.modules.uniswap.v2.constants import SWAP_SIGNATURE
+from rotkehlchen.chain.evm.decoding.interfaces import DecoderInterface
+from rotkehlchen.chain.evm.decoding.structures import ActionItem
+from rotkehlchen.chain.evm.decoding.uniswap_v2_like.common import (
+    compute_uniswap_v2_like_pool_address_with_unpacked_addresses,
+    decode_uniswap_like_deposit_and_withdrawals,
+    decode_uniswap_v2_like_swap,
+)
+from rotkehlchen.chain.evm.decoding.uniswap_v2_like.constants import (
+    BURN_SIGNATURES,
+    MINT_SIGNATURE,
+)
+from rotkehlchen.chain.evm.structures import EvmTxReceiptLog
+from rotkehlchen.logging import RotkehlchenLogsAdapter
+from rotkehlchen.types import ChecksumEvmAddress, EvmTransaction
+
+if TYPE_CHECKING:
+    from rotkehlchen.chain.evm.decoding.base import BaseDecoderTools
+    from rotkehlchen.chain.evm.node_inquirer import EvmNodeInquirer
+    from rotkehlchen.history.events.structures.evm_event import EvmEvent
+    from rotkehlchen.user_messages import MessagesAggregator
+
+
+logger = logging.getLogger(__name__)
+log = RotkehlchenLogsAdapter(logger)
+
+
+class UniswapV2LikeDecoder(DecoderInterface):
```

Comment:
it's not used, but it would make sense to use it for `chain/ethereum/modules/uniswap/v2/decoder.py` and `rotkehlchen/chain/ethereum/modules/sushiswap/decoder.py`

Though I am unsure if doing the work in the `post_decoding_rules` as I am doing here is better or worse than using normal decoding rules. It's just that for syncswap the order of the events (I think for the deposit/withdraw) was reversed so I needed to access the decoded transfer events at the beginning.

Reply:
I know that you can't always abstract similar logic and trying to fit everything in one box is not always feasible, and can lead to performance problems. However I also think that a good principle to strive for is to abstract as much as possible, so that the common logic for decoding swaps would work in principle for any AMM, so that we can add add many AMMs as possible using as little code as possible. In some cases (such as for any AMM based on Uniswap V2 which is very popular), we might want to resort to the common denominator (in this case using post-decoding) if that would work for most AMM contracts.

Reply:
post decoding rules is the worst type of rule for performance. I would not use it in an abstraction.